### PR TITLE
Remove the `rustc_private` feature

### PR DESCRIPTION
**Pull request by tomaka**
- Removes the `rustc_private` feature
 - Renames `deflate` to `inflate`
 - Adds `deflate` module
 - Adds `ZlibEncoder`
 - ~~Fixes a bug when reading uncompressed blocks in `inflate` (please review this, see `src/png/inflate.rs` in the second commit)~~ (extracted to #379)

The compressor doesn't actually compress anything. The deflate algorithm allows blocks of uncompressed data, so we just write everything in uncompressed data. This should of course be changed later.

### CODE REVIEW

**Comment by nwin** [comment]
Nice workaround. ;) If we are lucky  #376 and #300 have the same origin as the bug you found.

**Comment by tomaka** [comment]
I'll submit a different PR for the bugfix.

**Comment by bvssvni** [comment]
@tomaka What was the strange thing you mentioned?

File: src/png/deflate.rs [hunk 1]
```diff
@@ -1,431 +1,89 @@
...
-        self.h.byte_align();
+        assert!(self.buf.len() <= u16::max_value() as usize);
+        let len = self.buf.len() as u16;
+        let nlen = !len;    // one's complement
 
-        let len   = try!(self.h.receive(16));
-        let _nlen = try!(self.h.receive(16));
+        let writer = self.writer.as_mut().unwrap();
+        try!(writer.write_u8(if last { 0x29 } else { 0 }));       // bits and stuff
+                  // TODO: should be 1  ^, but a bug in the PNGDecoder can be bypassed with 0x29
```
Comment (tomaka) [inline src/png/deflate.rs]: @bvssvni This

**Comment by andersk** [comment]
`cargo bench` fails since you deleted `examples/fractal.png`.

**Comment by tomaka** [comment]
Oops, my bad.

**Comment by tomaka** [comment]
So.... let's merge this?

**Comment by nwin** [comment]
I just don’t understand your hack with the magic 0x29 does it produce malformed streams or what?

**Comment by tomaka** [comment]
According to the specs, the first bit indicates whether this is the last block, then the next two bits indicate the type of block (here we pass 0).
Then if the type of block is 0, the decoder must ignore the bits remaining in the current byte (so ignore the next 5 bits).

However image's PNGDecoder reads the next 3 bits and does a right shift depending on the value of these bits. This is clearly a bug in the decoder, but it's very specific so I guess there has to be a reason for it to do so.

To be compliant we must write `bXXXXX001` (where `X` can be anything), but in order for the PNGDecoder to decode it we must write `bXX101001`.

**Comment by nwin** [comment]
ok

**Comment by bvssvni** [comment]
@tomaka You are awesome!

**Pull request merged**